azext_iot/tests/digitaltwins/__init__.py

```diff
@@ -172,6 +183,92 @@ def get_available_region(self, capacity: int = 1, skip_regions: list = None) ->
             )
         )
 
+    def ensure_adx_resource(self):
+        """Ensure that the test has all ADX resources."""
+        # TODO: fix, see if we can use the kusto extension
+        if not settings.env.azext_dt_adx_cluster:
+            self.kwargs["cluster_body"] = json.dumps(
+                {
+                    "location": "eastus",
+                    "sku": {
+                        "name": "Dev(No SLA)_Standard_E2a_v4",
+                        "capacity": 1,
+                        "tier": "Basic"
+                    },
+                    "properties": {
+                        "enableDiskEncryption": False,
+                        "enableStreamingIngest": False,
+                        "enablePurge": False,
+                        "enableDoubleEncryption": False,
+                        "publicNetworkAccess": "Enabled",
+                        "engineType": "V3",
+                        "enableAutoStop": True
+                    }
+                }
+            )
+
+            cluster_create_state = self.embedded_cli.invoke(
+                "rest --method PUT --url 'https://management.azure.com/subscriptions/{}/resourceGroups/"
+                "{}/providers/Microsoft.Kusto/clusters/{}?api-version=2021-08-27' --body '{}'".format(
+                    self.get_subscription_id(),
+                    ADX_RG,
+                    ADX_CLUSTER,
+                    self.kwargs["cluster_body"],
+                )
+            ).as_json().get("properties").get("state")
+            retries = 0
+            while cluster_create_state not in ProvisioningStateType.FINISHED.value and retries < MAX_ADX_RETRIES:
+                sleep(300)
+                cluster_create_state = self.embedded_cli.invoke(
+                    "rest --method GET --url 'https://management.azure.com/subscriptions/{}/resourceGroups/"
+                    "{}/providers/Microsoft.Kusto/clusters/{}?api-version=2021-08-27'".format(
+                        self.get_subscription_id(),
+                        ADX_RG,
+                        ADX_CLUSTER,
+                    )
+                ).as_json().get("properties").get("state")
+                retries += 1
+
+            if retries == MAX_ADX_RETRIES:
+                logger.error("Waited 20 minutes")
+
+        if not settings.env.azext_dt_adx_database:
+            self.kwargs["database_body"] = json.dumps(
+                {
+                    "location": "eastus",
+                    "kind": "ReadWrite",
+                    "properties": {
+                        "softDeletePeriod": "P1D"
+                    }
+                }
+            )
+            database_creation_state = self.embedded_cli.invoke(
+                "rest --method PUT --url 'https://management.azure.com/subscriptions/{}/resourceGroups/"
```

Comment:
Just a note - these URLs are hard-coded for the default ARM endpoint `management.azure.com` but we might (possibly at some point) want to run tests against different control-plane endpoints.

Reply:
Will remove the ADX resource generation until the az kusto extension is no longer preview. This should solve the issue of these hardcoded urls

Reply:
Or you could drop the host, just like how the az rest commands in the implementation are (or should be) formatted -- in this case the CLI assumes the url as an Azure resource and prefixes the value with the ARM endpoint of the current cloud .